pkg/fileservice/remote_cache.go

```diff
@@ -0,0 +1,190 @@
+)
+
+// KeyRouter is an interface manages the remote cache information.
+type KeyRouter interface {
+	// Target returns the remote cache server service address of
+	// the cache key. If the cache do not exist in any node, it
+	// returns empty string.
+	Target(k CacheKey) string
+
+	// AddItem pushes a cache key item into a queue with a local
+	// cache server service address in the item. Gossip will take
+	// all the items and send them to other nodes in gossip cluster.
+	AddItem(key CacheKey, operation gpb.Operation)
+}
+
+type TargetCacheKeys map[string][]*pb.RequestCacheKey
+
+type KeyRouterFactory func() KeyRouter
+
+// RemoteCache is the cache for remote read.
+type RemoteCache struct {
+	// cacheClient sends the cache request to cache server.
+	cacheClient client.CacheClient
+	// keyRouterFactory is used to set the keyRouter async.
+	keyRouterFactory KeyRouterFactory
+	// keyRouter is used to find out which node we should send
+	// cache request to.
+	keyRouter KeyRouter
+}
+
+var _ IOVectorCache = new(RemoteCache)
+
+func NewRemoteCache(client client.CacheClient, factory KeyRouterFactory) *RemoteCache {
+	return &RemoteCache{
+		cacheClient:      client,
+		keyRouterFactory: factory,
+	}
+}
+
+func (r *RemoteCache) Read(ctx context.Context, vector *IOVector) error {
+	if r.keyRouterFactory == nil {
+		return nil
+	}
+	if r.keyRouter == nil {
+		r.keyRouter = r.keyRouterFactory()
+		if r.keyRouter == nil {
+			return nil
+		}
+	}
+
+	path, err := ParsePath(vector.FilePath)
+	if err != nil {
+		return err
+	}
+
+	targetCacheKeys := make(TargetCacheKeys, len(vector.Entries))
+	for i, entry := range vector.Entries {
+		if entry.done {
+			continue
+		}
+		if entry.Size < 0 {
+			continue
+		}
+
+		cacheKey := CacheKey{
+			Path:   path.File,
+			Offset: entry.Offset,
+			Sz:     entry.Size,
+		}
+		target := r.keyRouter.Target(cacheKey)
+		if len(target) == 0 {
+			continue
+		}
+
+		if _, ok := targetCacheKeys[target]; !ok {
+			targetCacheKeys[target] = make([]*pb.RequestCacheKey, 0)
+		}
+		targetCacheKeys[target] = append(targetCacheKeys[target], &pb.RequestCacheKey{
+			Index:    int32(i),
+			CacheKey: &cacheKey,
+		})
+	}
+
+	for target, key := range targetCacheKeys {
+		req := r.cacheClient.NewRequest(pb.CmdMethod_RemoteRead)
+		req.RemoteReadRequest = &pb.RemoteReadRequest{
+			RequestCacheKey: key,
+		}
+
+		if err := func(ctx context.Context) error {
+			ctx, cancel := context.WithTimeout(ctx, time.Second*3)
+			defer cancel()
+			resp, err := r.cacheClient.SendMessage(ctx, target, req)
+			if err != nil {
+				return err
+			}
+			defer r.cacheClient.Release(resp)
+			if resp.RemoteReadResponse != nil {
+				for _, cacheData := range resp.RemoteReadResponse.ResponseCacheData {
+					idx := int(cacheData.Index)
+					if idx < len(vector.Entries) {
+						// FIXME: how to check it is done? the data may not read from the
+						// cache, but from local storage.
+						if cacheData.Hit {
```

Comment:
I think this should be `if cacheData.Hit { entry.done = true; entry.CachedData = cacheData.Data }``. Caches only set the CachedData field, not the Data field. See the memory cache codes